aspnetcore/fundamentals/routing.md

```diff
@@ -459,172 +570,235 @@ Regular expressions used in routing often start with the caret (`^`) character a
-> * `IUrlHelper.Page` always overrides the current `handler` route value with `null` as an explicit values unless overridden by the user. 
->
-> Users are often surprised by the behavioral details of ambient values, because MVC doesn't seem to follow its own rules. For historical and compatibility reasons certain route values (`action`, `controller`, `page`, and `handler`) have their own special-case behavior.
->
-> The equivalent functionality provided by `LinkGenerator.GetPathByAction` and `LinkGenerator.GetPathByPage` duplicates these quirks of `IUrlHelper` for compatibility.
+[!code-csharp[](routing/samples/3.x/RoutingSample/Controllers/WidgetController.cs?name=snippet2)]
+
+The preceding  method returns `/Home/Subscribe/17`
+
+The following code in the `WidgetController` returns `/Widget/Subscribe/17`:
+
+[!code-csharp[](routing/samples/3.x/RoutingSample/Controllers/WidgetController.cs?name=snippet3)]
+
+The following code provides the controller from ambient values in the current request and explicit values: `{ action = "Edit", id = 17, }`:
+
+[!code-csharp[](routing/samples/3.x/RoutingSample/Controllers/GadgetController.cs?name=snippet)]
+
+In the preceding code:
+
+* `/Gadget/Edit/17` is returned.
+* <xref:Microsoft.AspNetCore.Mvc.ControllerBase.Url> gets the <xref:Microsoft.AspNetCore.Mvc.IUrlHelper>.
+* <xref:Microsoft.AspNetCore.Mvc.UrlHelperExtensions.Action*> 	
+generates a URL with an absolute path for an action method. The URL contains the specified `action` name and `route` values.
+
+The following code provides ambient values from the current request and explicit values: `{ page = "./Edit, id = 17, }`:
+
+[!code-csharp[](routing/samples/3.x/RoutingSample/Pages/Index.cshtml.cs?name=snippet)]
+
+The preceding code sets `url` to  `/Edit/17` when the Edit Razor Page contains the following page directive:
+
+ `@page "{id:int}"`
+
+If the Edit page doesn't contain the `"{id:int}"` route template, `url` is `/Edit?id=17`.
+
+The behavior of MVC's <xref:Microsoft.AspNetCore.Mvc.IUrlHelper> adds a layer of complexity in addition to the rules described here:
+
+* `IUrlHelper` always provides the route values from the current request as ambient values.
+* [IUrlHelper.Action](xref:Microsoft.AspNetCore.Mvc.UrlHelperExtensions.Action*) always copies the current `action` and `controller` route values as explicit values unless overridden by the developer.
+* [IUrlHelper.Page](xref:Microsoft.AspNetCore.Mvc.UrlHelperExtensions.Page*) always copies the current `page` route value as an explicit values unless overridden. <!--by the user-->.
+* `IUrlHelper.Page` always overrides the current `handler` route value with `null` as an explicit values unless overridden.
+
+Users are often surprised by the behavioral details of ambient values, because MVC doesn't seem to follow its own rules. For historical and compatibility reasons certain route values  such as `action`, `controller`, `page`, and `handler` have their own special-case behavior.
```

Comment:
```suggestion
Users are often surprised by the behavioral details of ambient values, because MVC doesn't seem to follow its own rules. For historical and compatibility reasons, certain route values  such as `action`, `controller`, `page`, and `handler` have their own special-case behavior.
```